### PR TITLE
Counter integration test

### PR DESCRIPTION
**Pull request by nfurfaro**
Adds a new integration test for the counter app. The difference from the pre-existing tests is that the new one simulates the dev user experience by using the counter's cli binary directly, and runs queries against a local builder instance.

### CODE REVIEW

File: apps/counter/app/tests/builder_tests.rs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+) -> u32 {
+    let increment_output = TokioCommand::new("cargo")
+        .args(&[
+            "run",
+            "--",
+            "increment-count",
+            &format!("http://{}", node_address.as_str()),
+            &format!("http://{}", builder_address.as_str()),
+            pint_directory,
+        ])
```
Comment (freesig) [inline apps/counter/app/tests/builder_tests.rs:92]:
```suggestion
        .args([
            "run",
            "--",
            "increment-count",
            &format!("http://{}", node_address.as_str()),
            &format!("http://{}", builder_address.as_str()),
            pint_directory,
        ])
```
not sure why CI isn't picking it up but clippy is complaining locally about this.

File: apps/counter/app/tests/builder_tests.rs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    (builder_process, node_address, builder_address)
+}
+
+async fn deploy_contract(builder_address: String) {
+    let deploy_output = TokioCommand::new("essential-rest-client")
+        .args(&[
+            "deploy-contract",
+            &format!("http://{}", builder_address.as_str()),
+            concat!(env!("CARGO_MANIFEST_DIR"), "/../pint/out/debug/pint.json").into(),
+        ])
```
Comment (freesig) [inline apps/counter/app/tests/builder_tests.rs:161]:
```suggestion
        .args([
            "deploy-contract",
            &format!("http://{}", builder_address.as_str()),
            concat!(env!("CARGO_MANIFEST_DIR"), "/../pint/out/debug/pint.json").into(),
        ])
```

File: apps/counter/app/tests/builder_tests.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    count
+}
+
+async fn increment_count(
+    node_address: String,
+    builder_address: String,
+    pint_directory: &str,
+) -> u32 {
+    let increment_output = TokioCommand::new("cargo")
+        .args(&[
```
Comment (freesig) [inline apps/counter/app/tests/builder_tests.rs:85]:
```suggestion
        .args([
```

File: apps/counter/app/tests/builder_tests.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        new_count = captures[1].parse::<u32>().expect("Failed to parse count");
+    }
+    new_count
+}
+
+async fn start_essential_builder() -> (Child, String, String) {
+    let mut builder_process = TokioCommand::new("essential-builder")
+        .arg("--block-interval-ms")
+        .arg("100")
+        // @todo can we remove ?
```
Comment (freesig) [inline apps/counter/app/tests/builder_tests.rs:117]: fix has landed but we need to publish a new builder and node first

File: apps/counter/app/tests/builder_tests.rs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+            pint_directory,
+        ])
+        .output()
+        .await
+        .expect("Failed to execute command");
+
+    // Read stdout
+    let stdout_str = String::from_utf8(increment_output.stdout).expect("Failed to parse stdout");
+
+    let _stderr_str = String::from_utf8(increment_output.stderr).expect("Failed to parse stderr");
```
Comment (freesig) [inline apps/counter/app/tests/builder_tests.rs:100]: Do we need this?

File: apps/counter/app/tests/builder_tests.rs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        ])
+        .output()
+        .await
+        .expect("Failed to execute command");
+
+    assert!(read_output.status.success(), "Command failed to run");
+
+    let stdout_str = String::from_utf8(read_output.stdout).expect("Failed to parse stdout");
+
+    let _stderr_str = String::from_utf8(read_output.stderr).expect("Failed to parse stderr");
```
Comment (freesig) [inline apps/counter/app/tests/builder_tests.rs:68]: Also this?

File: apps/counter/app/Cargo.toml [hunk 7]
```diff
@@ -11,6 +11,7 @@ essential-app-utils.workspace = true
 essential-hash.workspace = true
 essential-rest-client.workspace = true
 essential-types.workspace = true
+regex = "1.11.1"
```
Comment (freesig) [inline apps/counter/app/Cargo.toml:14]: Maybe just a dev dependencies?

**Review by freesig — APPROVED** [review]
Nice one @nfurfaro . Super clean code! Also happy first official contribution. Just a few minor suggestions.

File: pkgs/essential-minimal.nix [hunk 8]
```diff
@@ -1,12 +1,14 @@
...
 , pint
 , symlinkJoin
 }:
 symlinkJoin {
   name = "essential-minimal";
   paths = [
+    essential-rest-client
     essential-builder
     essential-node
     pint
```
Comment (mitchmindtree) [inline pkgs/essential-minimal.nix:14]: This is probably fine, but we should re-think the names of these `devShell`s and think about exactly what we want to expose at some point as "minimal" is probably a bit ambiguous and its not clear where to draw the line

File: apps/counter/app/tests/builder_tests.rs [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+
+#[tokio::test]
+async fn builder_integration() {
+    let (_builder_process, node_address, builder_address) = start_essential_builder().await;
+
+    let _ = compile_pint_project(concat!(env!("CARGO_MANIFEST_DIR"), "/../pint").into())
+        .await
+        .unwrap();
+    // ensure builder has started and compilation is done
+    sleep(Duration::from_secs(1)).await;
```
Comment (supiket) [inline apps/counter/app/tests/builder_tests.rs:20]: Do we need this sleep here?

**Review by mitchmindtree — APPROVED** [review]
Nice one @nfurfaro!! Exciting to see you land some Rust stuff :rocket:

**Review by supiket — APPROVED** [review]
Nice one @nfurfaro, so happy to the first contribution! 💮

File: apps/counter/app/tests/builder_tests.rs (hunk 9 above)
Comment (nfurfaro) [reply apps/counter/app/tests/builder_tests.rs:20]: good catch; not needed any more!

**Pull request merged**